pkg/listwatch/listwatch_test.go

```diff
@@ -30,69 +32,61 @@ import (
 
 var _ watch.Interface = &multiListerWatcher{}
 
-func setupMultiWatch(t *testing.T, namespaces []string, rvs map[string]string) (map[string]*watch.FakeWatcher, *multiListerWatcher) {
+func checkError(t *testing.T, msg string, err error, args ...interface{}) {
+	if err != nil {
+		t.Errorf(msg, args...)
+	}
+}
+
+//nolint:unparam
+func setupMultiWatch(t *testing.T, namespaces []string, exampleObject runtime.Object,
+	resyncPeriod time.Duration) (map[string]*watch.FakeWatcher, *sync.Mutex, *multiListerWatcher) {
 	n := len(namespaces)
 	ws := make(map[string]*watch.FakeWatcher, n)
+	wsLock := sync.Mutex{}
 
 	mlw := newMultiListerWatcher(namespaces, func(namespace string) cache.ListerWatcher {
 		return &cache.ListWatch{
 			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
 				l := metav1.List{}
-				l.ListMeta.ResourceVersion = rvs[namespace]
+				l.ListMeta.ResourceVersion = "0"
 				return &l, nil
 			},
 			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				wsLock.Lock()
+				defer wsLock.Unlock()
 				w := watch.NewFake()
 				ws[namespace] = w
 				return w, nil
 			},
 		}
-	})
+	}, exampleObject, resyncPeriod)
 
 	if _, err := mlw.List(metav1.ListOptions{}); err != nil {
-		t.Fatalf("failed to invoke List: %v", err)
+		t.Errorf("failed to invoke List: %v", err)
 	}
 
-	mlw.resourceVersions = rvs
-	if err := mlw.newMultiWatch(metav1.ListOptions{}); err != nil {
-		t.Fatalf("failed to create new multiWatch: %v", err)
+	if _, err := mlw.Watch(metav1.ListOptions{}); err != nil {
+		t.Errorf("failed to create new multiWatch: %v", err)
 	}
-	return ws, mlw
-}
-
-func TestNewMultiWatch(t *testing.T) {
-	func() {
-		defer func() {
-			if r := recover(); r != nil {
-				t.Errorf("newMultiWatch should not panic when number of resource versions is less than ListerWatchers; got: %v", r)
-			}
-		}()
-		// Create a multiWatch from 2 ListerWatchers but only pass 1 resource version.
-		_, _ = setupMultiWatch(t, []string{"1", "2"}, map[string]string{"1": "resource1"})
-	}()
-	func() {
-		defer func() {
-			if r := recover(); r != nil {
-				t.Errorf("newMultiWatch should not panic when number of resource versions matches ListerWatchers; got: %v", r)
-			}
-		}()
-		// Create a multiWatch from 2 ListerWatchers and pass 2 resource versions.
-		_, _ = setupMultiWatch(t, []string{"1", "2"}, map[string]string{"1": "resource1", "2": "resource2"})
-	}()
+	return ws, &wsLock, mlw
 }
 
 func TestMultiWatchResultChan(t *testing.T) {
```

Comment:
can we add a test case for the issue that this is meant to solve? and if not, an acceptance test that runs your reproducer?

Reply:
I'm not sure what you mean, this was a test which was no longer applicable so was removed

Reply:
no I mean we should add a test that verifies that this PR solves the performance issue

Reply:
We already have that being done in QE, Yuanlin has a reproducer matching the original issue.  This PR tidies up the behaviour of the caching and handles the thrashing with exceptions being propagated up to the aggregated cache, I can add a test for that specifically to test that.  We may still be throttled by the API server, which is separate from this and being investigated.

Reply:
I've added a test for the error condition which caused problems with the sync, it validates that only one namespace is impacted by the changes and doesn't force it on others